core/app.go

```diff
@@ -0,0 +1,342 @@
+package core
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/joyent/containerpilot/backends"
+	"github.com/joyent/containerpilot/config"
+	"github.com/joyent/containerpilot/discovery"
+	"github.com/joyent/containerpilot/services"
+	"github.com/joyent/containerpilot/tasks"
+	"github.com/joyent/containerpilot/telemetry"
+	"github.com/joyent/containerpilot/utils"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+var (
+	// Version is the version for this build, set at build time via LDFLAGS
+	Version string
+	// GitHash is the short-form commit hash of this build, set at build time
+	GitHash string
+)
+
+// App ...
+type App struct {
+	DiscoveryService discovery.DiscoveryService
+	Services         []*services.Service
+	Backends         []*backends.Backend
+	Tasks            []*tasks.Task
+	Telemetry        *telemetry.Telemetry
+	PreStartCmd      *exec.Cmd
+	PreStopCmd       *exec.Cmd
+	PostStopCmd      *exec.Cmd
+	Command          *exec.Cmd
+	StopTimeout      int
+	QuitChannels     []chan bool
+	maintModeLock    *sync.RWMutex
+	signalLock       *sync.RWMutex
+	paused           bool
+	ConfigFlag       string
+}
+
+// NewApp ...
```

Comment:
In `Reload()` it looks like we're copying all the content of the newly initialized app on top of the existing `App` struct so that our locks still work? That LGTM as a process but maybe `InitializeApp` and `LoadApp` (and probably `HandlePolling` and `InMaintenanceMode`?) should be non-exported and so that the only public interfaces are:

- `NewApp` (i.e. create from scratch only)
- `Reload`
- `Run`
- `ToggleMaintenanceMode`
- `Terminate`